src/tooltip.tsx

```diff
@@ -135,10 +136,77 @@ const Tooltip: React.FC<Props> = ({
+
+    const getPosition = React.useCallback(
+        (position: Position | undefined) =>
+            isTabletOrSmaller ? position || defaultPositionMobile : position || defaultPositionDesktop,
+        [isTabletOrSmaller]
+    );
+
+    const [tooltipClientRect, setTooltipClientRect] = React.useState<DOMRect | undefined>(undefined);
+
+    const tooltipBoundingClientRect = React.useCallback(() => {
+        if (!tooltipRef.current) return undefined;
+
+        return tooltipRef.current.getBoundingClientRect();
+    }, []);
+
+    const hasRightSpace = React.useCallback(() => {
+        if (!tooltipClientRect) return undefined;
+
+        return targetBoundingClientRect.current.right + tooltipClientRect.width > window.innerWidth;
+    }, [tooltipClientRect]);
+
+    const hasLeftSpace = React.useCallback(() => {
+        if (!tooltipClientRect) return undefined;
+
+        return targetBoundingClientRect.current.left < tooltipClientRect.width;
+    }, [tooltipClientRect]);
+
+    const validatePosition = React.useCallback(
+        (position: Position) => {
+            if (!tooltipClientRect) return position;
+
+            const validatePositionLeft = (position: Position) => {
+                const hasTopSpace = targetBoundingClientRect.current.top > tooltipClientRect.height;
+                return hasLeftSpace() ? (hasTopSpace ? 'top' : 'bottom') : position;
+            };
+
+            const alternativePosition = (position: Position) => {
+                return hasRightSpace() ? validatePositionLeft('left') : position;
+            };
+
+            const positionValidated = {
+                top:
+                    targetBoundingClientRect.current.top < tooltipClientRect.height
+                        ? alternativePosition('bottom')
+                        : alternativePosition(position),
+                right: alternativePosition(position),
+                left:
+                    targetBoundingClientRect.current.left < tooltipClientRect.width
+                        ? hasRightSpace()
+                            ? targetBoundingClientRect.current.top < tooltipClientRect.height
+                                ? 'bottom'
+                                : 'top'
+                            : 'right'
+                        : position,
+                bottom:
+                    targetBoundingClientRect.current.bottom + tooltipClientRect.height > window.innerHeight
+                        ? alternativePosition('top')
+                        : alternativePosition(position),
+            };
+
+            return positionValidated[position];
+        },
+        [hasLeftSpace, hasRightSpace, tooltipClientRect]
+    );
+
+    const [position, setPosition] = React.useState<Position | undefined>(rest.position);
 
-    const position = getPosition(rest.position);
+    React.useEffect(() => {
+        setPosition(validatePosition(getPosition(rest.position)));
+    }, [getPosition, rest.position, validatePosition]);
```

Comment:
this is a derived state, no need to store in a new state, you can use a regular variable:
```
const position = validatePosition(getPosition(rest.position));
```

This will simplify the code and avoid re-renders